src/zenml/integrations/modal/utils.py

```diff
@@ -0,0 +1,695 @@
+#  You may obtain a copy of the License at:
+#
+#       https://www.apache.org/licenses/LICENSE-2.0
+#
+#  Unless required by applicable law or agreed to in writing, software
+#  distributed under the License is distributed on an "AS IS" BASIS,
+#  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express
+#  or implied. See the License for the specific language governing
+#  permissions and limitations under the License.
+"""Shared utilities for Modal integration components."""
+
+import os
+import subprocess
+import threading
+import time
+from typing import Any, Dict, List, Optional, Tuple, Union
+
+try:
+    import modal
+except ImportError:
+    modal = None  # type: ignore
+
+from zenml.config import ResourceSettings
+from zenml.config.resource_settings import ByteUnit
+from zenml.enums import StackComponentType
+from zenml.logger import get_logger
+from zenml.stack import Stack, StackValidator
+
+logger = get_logger(__name__)
+
+# Common environment variable for Modal orchestrator run ID
+ENV_ZENML_MODAL_ORCHESTRATOR_RUN_ID = "ZENML_MODAL_ORCHESTRATOR_RUN_ID"
+
+
+class ModalLogStreamer:
+    """Stream logs from Modal CLI in a separate thread."""
+
+    def __init__(self, app_name: str, call_id: Optional[str], logger: Any):
+        """Initialize the log streamer.
+
+        Args:
+            app_name: Name of the Modal app to stream logs from.
+            call_id: Optional function call ID for filtering logs.
+            logger: Logger instance to use for output.
+        """
+        self.app_name = app_name
+        self.call_id = call_id
+        self.logger = logger
+        self.log_stream_active = threading.Event()
+        self.log_thread: Optional[threading.Thread] = None
+
+    def start(self) -> None:
+        """Start log streaming in a background thread."""
+        self.log_stream_active.set()
+        self.log_thread = threading.Thread(
+            target=self._stream_logs, daemon=True
+        )
+        self.log_thread.start()
+
+    def stop(self) -> None:
+        """Stop log streaming."""
+        self.log_stream_active.clear()
+        if self.log_thread:
+            # Give the log thread a moment to clean up
+            time.sleep(0.5)
+
+    def _stream_logs(self) -> None:
+        """Stream logs from Modal CLI."""
+        try:
+            # Use modal CLI to stream logs (automatically streams while app is active)
+            cmd = [
+                "modal",
+                "app",
+                "logs",
+                self.app_name,
+                "--timestamps",
+            ]
+            self.logger.debug(f"Starting log stream: {' '.join(cmd)}")
+
+            process = subprocess.Popen(
+                cmd,
+                stdout=subprocess.PIPE,
+                stderr=subprocess.STDOUT,
+                text=True,
+                bufsize=1,  # Line buffered
+                universal_newlines=True,
+            )
+
+            # Stream logs line by line
```

Comment:
I think it's probably better to use subproccess with context manager to avoid resource leak

```suggestion
            with subprocess.Popen(
                cmd,
                stdout=subprocess.PIPE,
                stderr=subprocess.STDOUT,
                text=True,
                bufsize=1,  # Line buffered
                universal_newlines=True,
            ) as process:
```